core/src/test/scala/quasar/plugin/postgres/CsvSupport.scala

```diff
@@ -86,6 +105,93 @@ trait CsvSupport {
   }
 
+  def columnsOf[
+    F[_]: Async, P <: Poly1, R <: HList, K <: HList, V <: HList, T <: HList, S <: HList](
+    events: Stream[F, UpsertEvent[R]],
+    renderRow: P,
+    idColumn: Column[ColumnType.Scalar])(
+    implicit
+    keys: Keys.Aux[R, K],
+    values: Values.Aux[R, V],
+    getTypes: Mapper.Aux[asColumnType.type, V, T],
+    ktl: ToList[K, String],
+    ttl: ToList[T, ColumnType.Scalar])
+    : Stream[F, List[Column[ColumnType.Scalar]]] = {
+    val go = events.pull.peek1 flatMap {
+      case Some((UpsertEvent.Create(records), _)) =>
+        records.headOption match {
+          case Some(r) =>
+            val rkeys = r.keys.toList
+            val rtypes = r.values.map(asColumnType).toList
+
+            val columns = rkeys.zip(rtypes).map((Column[ColumnType.Scalar] _).tupled)
+
+            Pull.output1(columns.filter(c => c =!= idColumn))
+
+          case _ => Pull.done
+        }
+
+      case _ => Pull.done
+    }
+
+    go.stream
+  }
+
+  def toUpsertCsvSink[F[_]: Async, P <: Poly1, R <: HList, K <: HList, V <: HList, T <: HList, S <: HList](
+      dst: ResourcePath,
+      sink: ResultSink.UpsertSink[F, ColumnType.Scalar],
+      idColumn: Column[ColumnType.Scalar],
+      writeMode: QWriteMode,
+      renderRow: P,
+      events: Stream[F, UpsertEvent[R]])(
+      implicit
+      keys: Keys.Aux[R, K],
+      values: Values.Aux[R, V],
+      getTypes: Mapper.Aux[asColumnType.type, V, T],
+      renderValues: Mapper.Aux[renderRow.type, V, S],
+      ktl: ToList[K, String],
+      stl: ToList[S, String],
+      ttl: ToList[T, ColumnType.Scalar])
+      : Stream[F, OffsetKey.Actual[String]] = {
+
+    val encoded: Stream[F, DataEvent[OffsetKey.Actual[String]]] = events flatMap {
+      case UpsertEvent.Create(records) => {
+        Stream.emits(records)
+          .covary[F]
+          .through(
+            encodeCsvRecordsToChunk[F, renderRow.type, R, V, S](renderRow))
+          .map(DataEvent.Create)
+      }
+
+      case UpsertEvent.Commit(s) =>
+        Stream(
+          DataEvent.Commit(OffsetKey.Actual.string(s)))
+
+      case UpsertEvent.Delete(Ids.StringIds(is)) =>
+        Stream(
+          DataEvent.Delete(IdBatch.Strings(is.toArray, is.length)))
+
+      case UpsertEvent.Delete(Ids.LongIds(is)) =>
+        Stream(
+          DataEvent.Delete(IdBatch.Longs(is.toArray, is.length)))
+    }
+
+    columnsOf(events, renderRow, idColumn).flatMap(cols =>
+      sink.consume.apply(
+        ResultSink.UpsertSink.Args(
+          dst, idColumn, cols, writeMode, encoded)))
+  }
+
+  def encodeCsvRecordsToChunk[F[_]: Async, P <: Poly1, R <: HList, V <: HList, S <: HList](
+      renderRow: P)(
+      implicit
+      values: Values.Aux[R, V],
+      render: Mapper.Aux[renderRow.type, V, S],
+      ltl: ToList[S, String])
+      : Pipe[F, R, Chunk[Byte]] =
+    encodeCsvRecords[F, P, R, V, S](renderRow).andThen(bytes =>
+      bytes.chunks.fold(Chunk.empty[Byte])((l, r) => Chunk.concatBytes(Seq(l, r))))
```

Comment:
Is there a better way to do this? i.e convert a `Stream[F, Byte]` into a `Stream[F, Chunk[Byte]]` with a single chunk emitting all the bytes.

Reply:
What you wrote reasonable to me. Something like
```scala
Stream.eval(bytes.chunks.compile.to(List).map(Chunk.concatBytes(_)))
```
is an alternative, but I'm not sure if it is any more performant (not that it really matters here) and it is definitely less straightforward.

Reply:
I think I'll keep it as is. Not very important anyway.